babel-plugin-transform-jsx/lib/index.js

```diff
@@ -154,21 +58,139 @@ export default function ({ types: t }) {
   }
 
   /* ==========================================================================
+   * Visitors
+   * ======================================================================= */
+
+  const visitJSXElement = (path, state) => {
+
+    const {
+      opts: {
+        useVariables = false
+      }
+    } = state
+
+    /* ==========================================================================
+     * Node Transformers
+     * ======================================================================= */
+
+    const JSXIdentifier = node => t.stringLiteral(node.name)
+
+    const JSXNamespacedName = node => t.stringLiteral(`${node.namespace.name}:${node.name.name}`)
+
+    const JSXMemberExpression = transformOnType({
+      JSXIdentifier: node => t.identifier(node.name),
+      JSXMemberExpression: node => (
+        t.memberExpression(
+          JSXMemberExpression(node.object),
+          JSXMemberExpression(node.property)
+        )
+      )
+    })
+
+    const JSXElementName = transformOnType({
+      JSXIdentifier: useVariables
+        ? node => path.scope.hasBinding(node.name) ? t.identifier(node.name) : JSXIdentifier(node)
```

Comment:
Great analysis, and you might actually be right about the ampersand thing 😊

My original idea was a little more ducktyp-ee. Have one configuration option named `variable` (to match the design of `module` and `function`, it could be `useVariable`). If the value were `true` we use the default regex (probably `^[A-Z]` as you suggest). If it's a string we use that as the regex. By default it would be `false`.